### PR TITLE
feat: add optional TLS communication for the metrics exporter on the instances

### LINKED ISSUE #4710: [Feature]: the metrics port on instance pods should support HTTPS
### Is there an existing issue already for this feature request/idea?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new feature request to be evaluated.

### What problem is this feature going to solve? Why should it be added?

As part of increased security for CloudNativePG, the prometheus metrics port
served in port 9187 should support TLS

### Describe the solution you'd like

Use basic auth and TLS which is supported by Prometheus.
https://prometheus.io/docs/prometheus/latest/configuration/https/

### Describe alternatives you've considered

none

### Additional context

_No response_

### Backport?

Yes

### Are you willing to actively contribute to this feature?

No

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by jsilvela**
Enable optional TLS communication on the Cluster's metrics port.
When TLS is enabled, the metrics exporter uses the same server certificate
used by PostgreSQL.

Closes #4710

### CODE REVIEW

File: pkg/specs/podmonitor.go [hunk 1]
```diff
@@ -41,8 +44,41 @@ func (c ClusterPodMonitorManager) BuildPodMonitor() *monitoringv1.PodMonitor {
...
+					},
+					Key: certs.CACertKey,
+				},
+			},
+			KeySecret: &corev1.SecretKeySelector{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: c.cluster.GetServerCASecretName(),
+				},
+				Key: certs.CAPrivateKeyKey,
+			},
```
Comment (jsilvela) [inline pkg/specs/podmonitor.go:73]: Per Marco, look into leaving empty

**Comment by jsilvela** [comment]
@gbartolini I was just looking at that. You're right.

**Comment by jsilvela** [comment]
After discussion with @mnencia we decided to leave this PR as is,
and create a new issue for the manager metrics.
https://github.com/cloudnative-pg/cloudnative-pg/issues/5070
cc/ @gbartolini

**Review by mnencia — APPROVED** [review]

**Review by armru — APPROVED** [review]

**Comment by mnencia** [comment]
While reviewing the documentation, I realized that our API here is incomplete. If someone is implementing TLS security in metrics scraping, they probably want more versatility in the TLS configuration, something like the following:

```
type TLSConfig struct {
	// Secret containing the CA cert to use to validate the client.
	CASecret *v1.SecretKeySelector `json:"caSecret,omitempty"`
	// Secret containing the server cert file.
	CertSecret *v1.SecretKeySelector `json:"certSecret,omitempty"`
	// Secret containing the server key file.
	KeySecret *v1.SecretKeySelector `json:"keySecret,omitempty"`
}
```

I don't say we should do it now, but we should be prepared to add it. So before merging, I want to check this idea and have a future-proof API.

**Comment by mnencia** [comment]
After a brief discussion, we concluded that the optimal interface to implement is as follows:

```yaml
spec:
  monitoring:
    tls:            
      enable: true 
```

This approach provides more room for future enhancements without cluttering the monitoring section.

**Review by mnencia — APPROVED** [review]

**Comment by mnencia** [comment]
Waiting for the E2E result before merging.

File: docs/src/monitoring.md [hunk 2]
```diff
@@ -90,6 +91,49 @@ spec:
...
 
+### Enabling TLS on the Metrics Port
+
+To enable TLS communication on the metrics port, configure the `spec.monitoring.tls.enable`
+setting to `true`. This setup ensures that the metrics exporter uses the same
+server certificates as PostgreSQL to secure communication on port 5432.
+
+!!! important
+    Changing the `spec.monitoring.tls.enable` value will cause all instances to restart.
+    Plan accordingly to minimize downtime.
```
Comment (jsilvela) [inline docs/src/monitoring.md:102]: "All instances to restart" seems very chaotic, where this is our well known operation of a rolling restart.
And "minimize downtime" ... with a well performed rolling restart, there would be none, right?

Comment (jsilvela) [inline docs/src/monitoring.md:102]:
```suggestion
    Changing the `spec.monitoring.tls.enable` value will initiate a rolling restart of the cluster.
```

Comment (mnencia) [reply docs/src/monitoring.md:102]: What about?

```suggestion
    Changing the `spec.monitoring.tls.enable` value will cause a rollout of all instances in the cluster.
```

File: config/crd/bases/postgresql.cnpg.io_clusters.yaml [hunk 3]
```diff
@@ -3909,6 +3909,18 @@ spec:
                           type: string
                       type: object
                     type: array
+                  tls:
+                    description: |-
+                      Configure TLS communication for the metrics endpoint.
+                      Changing tls.enable option will force a rollout of all instances.
+                    properties:
+                      enable:
```
Comment (gbartolini) [inline config/crd/bases/postgresql.cnpg.io_clusters.yaml:3917]: Why 'enable` when we already have `enabled` in other places?

Comment (mnencia) [reply config/crd/bases/postgresql.cnpg.io_clusters.yaml:3917]: Because you are right, and this is a typo, nobody noticed.

Comment (mnencia) [reply config/crd/bases/postgresql.cnpg.io_clusters.yaml]: fix committed

File: config/crd/bases/postgresql.cnpg.io_clusters.yaml [hunk 4]
```diff
@@ -3909,6 +3909,18 @@ spec:
                           type: string
                       type: object
                     type: array
+                  tls:
+                    description: |-
+                      Configure TLS communication for the metrics endpoint.
+                      Changing tls.enable option will force a rollout of all instances.
```
Comment (gbartolini) [inline config/crd/bases/postgresql.cnpg.io_clusters.yaml:3915]: `enable` -> `enabled`

Comment (mnencia) [reply config/crd/bases/postgresql.cnpg.io_clusters.yaml]: fixed

**Review by NiccoloFei — APPROVED** [review]

**Pull request merged**